mobile/library/common/internal_engine.cc

```diff
@@ -17,14 +17,61 @@
 namespace Envoy {
 namespace {
 constexpr absl::Duration ENGINE_RUNNING_TIMEOUT = absl::Seconds(30);
-// Google DNS address used for IPv6 probes.
-constexpr absl::string_view IPV6_PROBE_ADDRESS = "2001:4860:4860::8888";
-constexpr uint32_t IPV6_PROBE_PORT = 53;
 
 // There is only one shared MobileProcessWide instance for all Envoy Mobile engines.
 MobileProcessWide& initOnceMobileProcessWide(const OptionsImplBase& options) {
   MUTABLE_CONSTRUCT_ON_FIRST_USE(MobileProcessWide, options);
 }
+
+Network::Address::InstanceConstSharedPtr ipv6ProbeAddr() {
+  // Use Google DNS IPv6 address for IPv6 probes.
+  CONSTRUCT_ON_FIRST_USE(Network::Address::InstanceConstSharedPtr,
+                         new Network::Address::Ipv6Instance("2001:4860:4860::8888", 53));
+}
+
+Network::Address::InstanceConstSharedPtr ipv4ProbeAddr() {
+  // Use Google DNS IPv4 address for IPv4 probes.
+  CONSTRUCT_ON_FIRST_USE(Network::Address::InstanceConstSharedPtr,
+                         new Network::Address::Ipv4Instance("8.8.8.8", 53));
+}
+
+bool hasNetworkChanged(int network_type, int prev_network_type) {
+  // The network types are both cellular and don't differ in the Network::Generic (e.g. VPN) flag,
+  // so don't consider it a network change. If the previous and current network types differ, it's a
+  // difference in cellular network subtypes (e.g. 4G to 5G).
+  if ((network_type & static_cast<int>(NetworkType::WWAN)) &&
+      (prev_network_type & static_cast<int>(NetworkType::WWAN)) &&
+      (network_type & static_cast<int>(NetworkType::Generic)) ==
+          (prev_network_type & static_cast<int>(NetworkType::Generic))) {
+    return false;
+  }
+  // Otherwise, if the current network type is different from the previous network type, it's a
+  // network change.
+  return network_type != prev_network_type;
+}
+
+bool areIpAddressesDifferent(const Network::Address::InstanceConstSharedPtr& addr,
+                             const Network::Address::InstanceConstSharedPtr& prev_addr) {
+  if (addr == nullptr && prev_addr == nullptr) {
+    return false;
+  }
+  if (addr == nullptr || prev_addr == nullptr) {
+    // Change in presence of an IP address is treated as a difference.
+    return true;
+  }
+  if (addr->ip()->version() != prev_addr->ip()->version()) {
+    // IP address versions are different.
+    return true;
+  }
+  if (addr->ip()->version() == Network::Address::IpVersion::v6) {
+    // Both are IPv6 but the addresses are different.
+    return addr->ip()->ipv6()->address() != prev_addr->ip()->ipv6()->address();
+  } else {
+    // Both are IPv4 but the addresses are different.
+    return addr->ip()->ipv4()->address() != prev_addr->ip()->ipv4()->address();
+  }
+  return false;
```

Comment:
nit: The `else` can be removed and can be simplified to simply

```
// Both are IPv4 but the addresses are different.
return addr->ip()->ipv4()->address() != prev_addr->ip()->ipv4()->address();
```

Reply:
done